Docker HTTPS Connection for Local Testing

JonahZW opened the pull request:
Adds HTTPS support to the Docker setup for the application by generating self-signed certificates if they do not already exist. It also ensures that the generated certificates include the necessary Subject Alternative Names (SANs) to be trusted by modern browsers.

This better facilitates local testing by properly mimicking the live production environment and enabling SSL connections.

JonahZW committed:
Enable SSL Connection for Docker

Creates a basic certificate and uses it in the Docker instance to enable SSL connections and use https instead of http

JonahZW committed:
Create own SSL key, add to keystore

Automatically generates individual key and certificate when running ./start.sh and adds it to trusted keystore

JonahZW committed:
Update .gitignore

dlobron commented on ChugBotWithTestDBData.sql:
```diff
@@ -271,33 +271,6 @@ INSERT INTO `category_tables` VALUES ('blocks',1,0),('bunks',2,1),('campers',3,1
-/*!40000 ALTER TABLE `chug_dedup_instances` DISABLE KEYS */;
-INSERT INTO `chug_dedup_instances` VALUES ('Cooking','Cooking'),('Dance','Dance'),('Flag Football','Flag Football'),('Glass Painting','Glass Painting'),('Jewelry-Making','Jewelry-Making'),('Martial Arts','Martial Arts'),('Melty Beads','Melty Beads'),('Pickleball','Pickleball'),('Sand Art','Sand Art'),('Soccer','Soccer'),('Softball','Softball'),('Teva','Teva'),('Ultimate Frisbee','Ultimate Frisbee ...
-/*!40000 ALTER TABLE `chug_dedup_instances` ENABLE KEYS */;
-UNLOCK TABLES;
```
Just checking that this deletion was intentional?

JonahZW replied:
Correct - it was throwing errors trying to import the db when this was included, but as far as I can tell through the rest of the codebase, this dedup table was deprecated and no longer being used, in favor of one which tracks the relationship by `chug_id` instead of `name`

dlobron commented on Docker/my-httpd.conf:
```diff
@@ -0,0 +1,27 @@
+LoadModule ssl_module modules/mod_ssl.so
+
+# Listening on port 443 for HTTPS
+#Listen 443
```
Should we uncomment this to listen on 443?

JonahZW replied:
Oops, yeah, missed that... it does seem to work without that being enabled, though, which is strange. I'll fix that :)

dlobron commented on start.sh:
```diff
@@ -2,14 +2,52 @@
 
+# Wait for the MySQL container to be fully up
+echo "Waiting for MySQL to be ready..."
+until docker exec db mysqladmin ping -h "db" --silent; do
```
nit: We could optionally add a counter and max_retries variable, so we exit if MySQL is not responding. I don't think it's a necessity for this PR, unless this script could potentially run remotely.

JonahZW replied:
I'm 99.5% sure it can only be run locally for our testing instance - including a counter is a good idea, I could do that, but when I've been testing it usually loads the db within ~7 tries, which is on par with how it did previously

dlobron reviewed (commented):
This is great! Just to confirm, does it do anything beyond adding encryption for the local Docker container? I'm just curious if it will also affect the AWS instances.

JonahZW commented:
Yes, it only affects the local Docker container - none of the files changed have are dependancies of the files which run on the AWS instances, nor are any of them included in that zip.

JonahZW committed:
Script to easily pull up testing database

Creates a script which can be called to easily pull up the active mysql database (without searching for the container id or selecting "camprama_chugbot_db" as the database to work in). When the Docker container isn't running, it automatically sends an error and quits.

Also fixes a minor bug in the docker https config

JonahZW commented:
I also just wrote a short script to easily start using the correct database container instance and the correct container in the MYSQL Interactive Prompt (https://github.com/JonahZW/ChugBot/commit/b24e3e726a09b2f40b05d864fbc3fdea58e461ed)

dlobron commented:
Great job! Thank you for answering all the questions! This will be very useful!
